daemons/controld/controld_execd.c

```diff
@@ -1978,7 +1978,41 @@ construct_op(lrm_state_t * lrm_state, xmlNode * rsc_op, const char *rsc_id,
             resolve_versioned_attrs(lrm_state, rsc, rsc_op, params);
         }
 #endif
+
+        /* pcmk_monitor_timeout overrides CRM_meta_timeout for stonith
+        * start/monitor
+        */
+        if (safe_str_eq(rsc->standard, PCMK_RESOURCE_CLASS_STONITH)
+            && pcmk__str_any_of(op->op_type, CRMD_ACTION_START,
+                                CRMD_ACTION_STATUS, NULL)) {
+
+            const char *st_mon_timeout = g_hash_table_lookup(params,
+                    "pcmk_monitor_timeout");
+
+            if (st_mon_timeout) {
+                op->timeout = crm_get_msec(st_mon_timeout);
+            }
+        }
+
         lrmd_free_rsc_info(rsc);
+    } else {
+        xmlNode *rsc = first_named_child(rsc_op, XML_CIB_TAG_RESOURCE);
+
+        if (rsc) {
+            const char *class = crm_element_value(rsc, XML_AGENT_ATTR_CLASS);
+
+            if (safe_str_eq(class, PCMK_RESOURCE_CLASS_STONITH)
+                && pcmk__str_any_of(op->op_type, CRMD_ACTION_START,
+                                    CRMD_ACTION_STATUS, NULL)) {
```

Comment:
FYI, lrm_state 'should' always be non-NULL when calling construct_op(), but something like this is always helpful to recover gracefully from a bug elsewhere.